### DISCUSSION TITLE
Is there any way we can have maxItemCount for HorizontalAxis?

### DISCUSSION THREAD

**Discussion by aaaleem** [patrykandpatrick/vico#464] [Questions]
"Is there a way to address the HorizontalAxis, similar to how we manage the VerticalAxis using:

`itemPlacer = AxisItemPlacer.Vertical.default(4)`

I am attempting to resolve the following issue:

Suppose I have data on the x-Axis organized by Year-Month, for example: 2020-10, 2020-11, 2020-12, 2021-1 ... 2023-08 (which i convert into Float for x-axis). In total, there are 37 values, and when using AxisValueFormatter<AxisPosition.Horizontal.Bottom>, I obtain a total of chartValues.getMaxMajorEntryCount() equals 39.

However, I want to limit the displayed values on the x-axis to only show the years once (2020, 2021, 2022, and 2023).

I am aware that I can utilize AxisItemPlacer.Horizontal.default() for spacing or xStep to skip steps in the x-axis. However, I'm unsure about the appropriate value to use to display years only once. I have experimented with various values, but the solution has not worked as expected. Could someone please assist me with this?"

**Comment by patrickmichalik** [comment]
Hello! Just to confirm, are the _x_ values Unix timestamps?

**Comment by aaaleem** [comment]
@patrickmichalik - Thanks for your reply. I calculate x values using this formula. 

```
/**
 * Converts Year and Month to decimal data
 * e.g
 * Jan 2020 would be 2020.0
 * Dec 2020 would be 2020.92
 *
 * Note: limits conversion to 2 decimal digits
 */
fun YearMonth.xToYearMonthDecimalData(): Float {
    return (this.year.toDouble() + (this.monthValue - 1) / 12.toDouble())
        .toBigDecimal()
        .setScale(2, RoundingMode.HALF_UP)
        .toFloat()
}
```

so everything is working fine however I want to display year value once in x-axis.

**Reply by patrickmichalik** [reply]
Got it. Two more questions:

1. Is your chart a column chart or a line chart?
2. How would you like the _x_-axis labels and ticks to be positioned? (Should a given year’s label align with January 1, or should it be in the middle, between June and July? In the latter case, should the ticks be aligned with the labels, or should they come between December 31 and January 1, separating neighboring years?)

**Reply by aaaleem** [reply]
Thank you for getting back to me. 

1. Line Chart
2. I want ticks to be positioned monthly (I mean the line chart should follow monthly data per year)  and the x-axis to be labeled as years like 2020, 2021, and 2022. Is it possible?

**Reply by patrickmichalik** [reply]
Thanks, @aaaleem! Does any of the following charts match what you have in mind? If not, could you please pick the one that’s the closest to the desired result and let me know what changes are needed? Please disregard the differences in _y_ values—they’re randomly generated. The first entry is for January of 2022.

<img src="https://github.com/patrykandpatrick/vico/assets/120058021/f6a6c07c-6c5e-4aa7-9dea-d0028f83936b" width="720" />
<img src="https://github.com/patrykandpatrick/vico/assets/120058021/5c2a68c9-dbb3-4941-9cb4-9812d563fa28" width="720" />
<img src="https://github.com/patrykandpatrick/vico/assets/120058021/881ccb0b-eb52-4eeb-a165-908ace5e029b" width="720" />

**Reply by aaaleem** [reply]
@patrickmichalik - Thank you for your quick response. 

The first and third charts are closer to what I have in my mind. Can you please let me know how should I update my x-axis?

**Reply by aaaleem** [reply]
@patrickmichalik - Any updates on this?

**Reply by patrickmichalik** [reply]
Sorry for the wait, @aaaleem. First, the manner in which the _x_ values are determined should be updated such that the spacing is even. Consider January to March of 2022. We have 2022, 2022.08, and 2022.17. The difference between the first value and the second one is 0.08, but the difference between the second value and the third one is 0.09, even though the three months should be spaced out evenly on the chart. Once this has been addressed, you can create a custom `AxisItemPlacer.Horizontal` implementation to position the ticks and labels as desired. Please see the `AxisItemPlacer.Horizontal` functions’ documentation for information on what they’re supposed to return. `getLabelValues` and `getLineValues` are the most important. For the remaining functions, it might be helpful to take a look at [`DefaultHorizontalAxisItemPlacer`](https://github.com/patrykandpatrick/vico/blob/master/vico/core/src/main/java/com/patrykandpatrick/vico/core/axis/horizontal/DefaultHorizontalAxisItemPlacer.kt). The default parameter values are [here](https://github.com/patrykandpatrick/vico/blob/master/vico/core/src/main/java/com/patrykandpatrick/vico/core/axis/AxisItemPlacer.kt#L119-L122). The `getLabelValues` and `getLineValues` implementations in `DefaultHorizontalAxisItemPlacer` are quite complex because they’re designed to handle all cases and are quite heavily optimized. In chart-specific `AxisItemPlacer.Horizontal` implementations, these functions can be significantly simpler. The idea is simply to return a list of label _x_ values and a list of line _x_ values.

**Reply by aaaleem** [reply]
@patrickmichalik - Thank you so much! I appreciate it. I am going to give it a try. 🙌

**Reply by patrickmichalik** [reply]
You’re welcome! Let me know if you have any questions.
